From 7bb9d0e5165ecf5cd53cd1076abc12f516053adb Mon Sep 17 00:00:00 2001
From: Yuxuan Shui <yshui@codeweavers.com>
Date: Mon, 6 Nov 2023 16:05:38 +0000
Subject: [PATCH 49/51] Revert ntdll: Tweak the binary representation of
 SRWLOCK.

There are applications that uses SRWLOCK in an invalid way and then checks its binary
representation. Tweak our representation a bit so they are happy.
---
 dlls/kernel32/tests/sync.c | 17 -----------------
 dlls/ntdll/sync.c          | 24 ++++++++++--------------
 2 files changed, 10 insertions(+), 31 deletions(-)

diff --git a/dlls/kernel32/tests/sync.c b/dlls/kernel32/tests/sync.c
index 3b584875fe7..ffe4e49fa50 100644
--- a/dlls/kernel32/tests/sync.c
+++ b/dlls/kernel32/tests/sync.c
@@ -2700,22 +2700,6 @@ static void test_srwlock_example(void)
     trace("number of total exclusive accesses is %ld\n", srwlock_protected_value);
 }
 
-static void test_srwlock_quirk(void)
-{
-    union { SRWLOCK *s; LONG *l; } u = { &srwlock_example };
-
-    if (!pInitializeSRWLock) {
-        /* function is not yet in XP, only in newer Windows */
-        win_skip("no srw lock support.\n");
-        return;
-    }
-
-    /* WeCom 4.x checks releasing a lock with value 0x1 results in it becoming 0x0. */
-    *u.l = 1;
-    pReleaseSRWLockExclusive(&srwlock_example);
-    ok(*u.l == 0, "expected 0x0, got %lx\n", *u.l);
-}
-
 static DWORD WINAPI alertable_wait_thread(void *param)
 {
     HANDLE *semaphores = param;
@@ -3427,7 +3411,6 @@ START_TEST(sync)
     test_condvars_base(&unaligned_cv.cv);
     test_condvars_consumer_producer();
     test_srwlock_base(&aligned_srwlock);
-    test_srwlock_quirk();
 #if defined(__i386__) || defined(__x86_64__)
     /* unaligned locks only work on x86 platforms */
     test_srwlock_base(&unaligned_srwlock.lock);
diff --git a/dlls/ntdll/sync.c b/dlls/ntdll/sync.c
index 7c759a81ab2..c1778c1d44e 100644
--- a/dlls/ntdll/sync.c
+++ b/dlls/ntdll/sync.c
@@ -463,10 +463,9 @@ DWORD WINAPI RtlRunOnceExecuteOnce( RTL_RUN_ONCE *once, PRTL_RUN_ONCE_INIT_FN fu
 
 struct srw_lock
 {
-    /* bit 0 - if the lock is held exclusive. bit 1.. - number of exclusive waiters. */
     short exclusive_waiters;
 
-    /* Number of owners.
+    /* Number of shared owners, or -1 if owned exclusive.
      *
      * Sadly Windows has no equivalent to FUTEX_WAIT_BITSET, so in order to wake
      * up *only* exclusive or *only* shared waiters (and thus avoid spurious
@@ -477,7 +476,7 @@ struct srw_lock
      * the latter waits only on the "owners" member. Note then that "owners"
      * must not be the first element in the structure.
      */
-    unsigned short owners;
+    short owners;
 };
 C_ASSERT( sizeof(struct srw_lock) == 4 );
 
@@ -508,7 +507,7 @@ void WINAPI RtlAcquireSRWLockExclusive( RTL_SRWLOCK *lock )
 {
     union { RTL_SRWLOCK *rtl; struct srw_lock *s; LONG *l; } u = { lock };
 
-    InterlockedExchangeAdd16( &u.s->exclusive_waiters, 2 );
+    InterlockedIncrement16( &u.s->exclusive_waiters );
 
     for (;;)
     {
@@ -523,9 +522,8 @@ void WINAPI RtlAcquireSRWLockExclusive( RTL_SRWLOCK *lock )
             if (!old.s.owners)
             {
                 /* Not locked exclusive or shared. We can try to grab it. */
-                new.s.owners = 1;
-                new.s.exclusive_waiters -= 2;
-                new.s.exclusive_waiters |= 1;
+                new.s.owners = -1;
+                --new.s.exclusive_waiters;
                 wait = FALSE;
             }
             else
@@ -560,7 +558,7 @@ void WINAPI RtlAcquireSRWLockShared( RTL_SRWLOCK *lock )
             old.s = *u.s;
             new = old;
 
-            if (!old.s.exclusive_waiters)
+            if (old.s.owners != -1 && !old.s.exclusive_waiters)
             {
                 /* Not locked exclusive, and no exclusive waiters.
                  * We can try to grab it. */
@@ -591,10 +589,9 @@ void WINAPI RtlReleaseSRWLockExclusive( RTL_SRWLOCK *lock )
         old.s = *u.s;
         new = old;
 
-        if (!(old.s.exclusive_waiters & 1)) ERR("Lock %p is not owned exclusive!\n", lock);
+        if (old.s.owners != -1) ERR("Lock %p is not owned exclusive!\n", lock);
 
         new.s.owners = 0;
-        new.s.exclusive_waiters &= ~1;
     } while (InterlockedCompareExchange( u.l, new.l, old.l ) != old.l);
 
     if (new.s.exclusive_waiters)
@@ -616,7 +613,7 @@ void WINAPI RtlReleaseSRWLockShared( RTL_SRWLOCK *lock )
         old.s = *u.s;
         new = old;
 
-        if (old.s.exclusive_waiters & 1) ERR("Lock %p is owned exclusive!\n", lock);
+        if (old.s.owners == -1) ERR("Lock %p is owned exclusive!\n", lock);
         else if (!old.s.owners) ERR("Lock %p is not owned shared!\n", lock);
 
         --new.s.owners;
@@ -647,8 +644,7 @@ BOOLEAN WINAPI RtlTryAcquireSRWLockExclusive( RTL_SRWLOCK *lock )
         if (!old.s.owners)
         {
             /* Not locked exclusive or shared. We can try to grab it. */
-            new.s.owners = 1;
-            new.s.exclusive_waiters |= 1;
+            new.s.owners = -1;
             ret = TRUE;
         }
         else
@@ -674,7 +670,7 @@ BOOLEAN WINAPI RtlTryAcquireSRWLockShared( RTL_SRWLOCK *lock )
         old.s = *u.s;
         new.s = old.s;
 
-        if (!old.s.exclusive_waiters)
+        if (old.s.owners != -1 && !old.s.exclusive_waiters)
         {
             /* Not locked exclusive, and no exclusive waiters.
              * We can try to grab it. */
-- 
2.53.0

